pkg/sign/signer.go

```diff
@@ -15,29 +15,218 @@
 package sign
 
 import (
+	"bytes"
 	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
 	"crypto/rand"
 	"crypto/sha256"
 	_ "crypto/sha512" // if user chooses SHA2-384 or SHA2-512 for hash
 	"crypto/x509"
 	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
 	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
 
 	protobundle "github.com/sigstore/protobuf-specs/gen/pb-go/bundle/v1"
 	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
 )
 
 type Signer interface {
-	Sign(data []byte) (*protobundle.Bundle, error)
+	Sign(content Content) (*protobundle.Bundle, error)
 }
 
-// TODO: implement
-// type Fulcio struct {
-//	baseURL       string
-//	identityToken string
-// }
+type Fulcio struct {
+	options *FulcioOptions
+}
+
+type FulcioOptions struct {
+	// URL of Fulcio instance
+	BaseURL string
+	// OIDC token to send to Fulcio instance
+	IdentityToken string
+	// Optional timeout for network requests
+	Timeout time.Duration
+	// Optional version string for user agent
+	LibraryVersion string
+}
+
+func SignerFulcio(opts *FulcioOptions) *Fulcio {
+	return &Fulcio{options: opts}
+}
+
+type jsonWebToken struct {
+	Sub string `json:"sub"`
+}
+
+type fulcioCertRequest struct {
+	Credentials      identityToken    `json:"credentials"`
+	PublicKeyRequest publicKeyRequest `json:"publicKeyRequest"`
+}
+
+type identityToken struct {
+	OIDCIdentityToken string `json:"oidcIdentityToken"`
+}
+
+type publicKeyRequest struct {
+	PublicKey         publicKey `json:"publicKey"`
+	ProofOfPossession string    `json:"proofOfPossession"`
+}
+
+type publicKey struct {
+	Algorithm string `json:"algorithm"`
+	Content   string `json:"content"`
+}
+
+type fulcioResponse struct {
+	SCT signedCertificateEmbeddedSct `json:"signedCertificateEmbeddedSct"`
+}
+
+type signedCertificateEmbeddedSct struct {
+	Chain chain `json:"chain"`
+}
+
+type chain struct {
+	Certificates []string `json:"certificates"`
+}
+
+func (f *Fulcio) Sign(content Content) (*protobundle.Bundle, error) {
+	// Get JWT from identity token
+	tokenParts := strings.Split(f.options.IdentityToken, ".")
+	if len(tokenParts) < 2 {
+		return nil, errors.New("Unable to get subject from identity token")
+	}
+
+	jwtString, err := base64.RawURLEncoding.DecodeString(tokenParts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	var jwt jsonWebToken
+	err = json.Unmarshal([]byte(jwtString), &jwt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate ephemeral keypair
```

Comment:
I don't want to couple artifact signing with requesting Fulcio certificate requests, as this would mean that we wouldn't support self-managed keys and identity certificates.

I would propose that the Sign interface be responsible for signing `content` and producing a bundle, and optionally fetching a certificate. The implementations of that interface could include:
* `EphemeralSigner`, which initializes a key in memory (This is the only one we need for now)
* `FileSigner` (name TBD) which reads and uses a key from disk 
* `KMSSigner`, which reads and uses a key from KMS

Each signer can take in an optional `Fulcio` which can retrieve a certificate for the key material and include it in the bundle. How would this look in practice? Maybe the `Sign` method could instead be `Sign(content Content, opts Options)`? An`Options` struct could include:
* Signature key scheme and hash (from https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_common.proto#L62) (with a default of `PKIX_ECDSA_P256_SHA_256`)
* Optional `Fulcio{}`
* Future: Optional `Rekor{}`
* Maybe: Optional `TimestampAuthority`?

A future improvement could be to have `Fulcio` implement an interface like `CertificateRequest` so you could bring your own CA too. If it's not much work, I might suggest doing that now too if you like that approach.

cc @kommendorkapten @codysoyland @phillmv for your thoughts!

Reply:
Separating out the key management part from Fulcio is a good idea. I could envision something like this:

```go
func NewFulcioSigner (s crypto.Signer, url string) *FulcioSigner {...} // implements RawSigner

type RawSigner {
    crypto.Signer
    AddVerificationMaterial(pb *protobundle.Bundle)
}
```

Then we can have methods for preparing `RawSigner`s for various key backend such as kms, file or ephemeral. We can add a new layer that wraps a `RawSigner` for embedding public key or cert into the verification material too to avoid repetitive code.
The we create one for DSSE and one for message digest that only takes in a `crypto.Signer` and options (witnesses). I think that could simplify the code a bit as then TSA client would only do one thing, countersign without needing to bother where to get the signature to counter sign, and where to store it? Now that code is in `tsa` module.

Something like (not valid code but I think the idea of wrapping makes it somewhat clear):
```go

type CertSigner struct {
    s crypto.Signer
    cert *x509.Certificate
}

func (s *CertSigner) AddVerificationMaterial(b *protobundle.Bundle) {
    bundle.VerificationMaterial: &protobundle.VerificationMaterial{
			Content: &protobundle.VerificationMaterial_X509Cert{
                             cert
        }
}

type FulcioSigner struct
    CertSigner
    url string
}

func (f *FulcioSigner) GetCert(idToken string) {
        f.cert = f.getCert(url, f.s.PublicKey(), idToken)
}

func (f *FulcioSigner) Sign(...) {
    return f.s.Sign(...)
}

type DsseSigner struct {
    s RawSigner
}

func (d* DsseSigner) Sign(content, opts) error {
    bundle := &protobundle.Bundle{...}   
    sig := d.s.Sign(content.PreAuthEncode(), ...)
    d.s.AddVerificationMaterial(bundle)
    env := Dsse.Envelope{}
    // add content to envelope
    counterSignatures := [][]byte   
    for _, s := range env.Signatures {
        for _, w := range opts.witness {
            cs := w.CounterSign(s)
            bundle.VerificationMaterial.TimestampVerificationData = append(counterSignatures, newTimestampVerificationData(cs))
        }
    }
    bundle.Content = &protobundle.Bundle_DsseEnvelope{...}
}
```

so the idea is that we can compose the final signer with either DSSE/message signature with either key or cert signer, with the different key implementations from kms, file or ephemeral. Fulcio would thus be a super type of the CertSigner to manage to extract the cert (_any_ type of private key material). The client would only need to call the GetCert function after the public key is known, but before it starts to sign.